In [41]:
import pandas as pd
import plotly.graph_objs as go
from plotly.offline import plot
from sklearn.ensemble import VotingClassifier,RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier,ExtraTreeClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegressionCV,LogisticRegression
from sklearn.naive_bayes import GaussianNB,BernoulliNB
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.preprocessing import StandardScaler
# from sk

In [2]:
data_train = pd.read_csv("Train.csv")

/home/whirldata/py35/lib/python3.5/site-packages/IPython/core/interactiveshell.py:3044: DtypeWarning:

Columns (746,835) have mixed types. Specify dtype option on import or set low_memory=False.



In [3]:
cols_val_counts = data_train.count().values
cols_names = data_train.count().index

In [34]:
cols_val_counts

array([17521, 17521, 17521, ..., 16347, 16347, 16335])

In [4]:
# cols_under_1000 = []
# cols_under_10 = []
# cols_btw_1000_2000 = []
# cols_under_10000 = []
cols_under_org_count = []
cols_without_na = []
for i,count in enumerate(cols_val_counts):
    if count < 17520:
        cols_under_org_count.append(cols_names[i])
    else:
        cols_without_na.append(cols_names[i])
        
#     if count < 1000:
# #         print(cols_names[i],"**********",count)
#         cols_under_1000.append(cols_names[i])
#     if count < 10000:
#         cols_under_10000.append(cols_names[i])
# #         print(cols_names[i],"**********",count)
#     if count > 1000 and count < 2000 :
#         cols_btw_1000_2000.append(cols_names[i])
#     if count < 10:
# #         print(cols_names[i],"**********",count)
#         cols_under_10.append(cols_names[i])

In [6]:
len(cols_under_org_count)

759

In [17]:
%%time
values_dict= {}
no_of_ones_in_withNA = {}
cols_ones_count = {}

for colss in cols_under_org_count:
    count_ones = 0
    cols__count = list(data_train[(~data_train[colss].isna()) & (data_train["Col2"]==1)]["Col2"])
    cols_ones_count.update({
        colss : len(cols__count)
        })
#     values_dict.update({
#         colss: cols2_count
#     })
    
#     ones_count = len(data_train[(data_train["Col463"].isna()) & (data_train["Col2"]==1)])
#     no_of_ones_in_withNA.update({
#         colss : ones_count
#     })

CPU times: user 13.7 s, sys: 563 µs, total: 13.7 s
Wall time: 13.7 s


In [20]:
from collections import defaultdict

group_based_ones = defaultdict(list)

for key, value in sorted(cols_ones_count.items()):
    group_based_ones[value].append(key)

In [35]:
y_ones = sorted([len(x) for x in group_based_ones.values()])

In [20]:
data_train[~data_train["Col463"].isna()]
data_train[~data_train["Col266"].isna()]
data_train[~data_train["Col288"].isna()]
data_train[~data_train["Col485"].isna()]
data_train[~data_train["Col724"].isna()]
data_train[~data_train["Col813"].isna()]
#In the below output we can find only 2 "1" class and 5 "0" class
#so here col463 is not important

Col1  Col2          Col3          Col4          Col5          Col6  \
123    92VUOOQ5     0   3696.020000   4144.623333   4144.623333   8289.246667   
264    AYEUUCK9     0  10395.423330  10820.044000   9016.703333   9016.703333   
3151   7NYWGOSR     1   1371.333333    543.169167    543.169167   1086.338333   
5926   O9LRG15E     0   7797.460000   4441.111111   3997.000000   6661.666667   
7002   O9GO5SQG     0  41676.546670  17637.871000  17569.688180  29396.451670   
8327   O9YUQMZR     1   8360.530000   6759.134444   8397.253000   8111.680000   
11025  O75KO0KE     0   2163.123333   4702.136000   5497.718333   5497.718333   

               Col7          Col8          Col9         Col10  ...  Col2388  \
123     4576.208333   9152.416667   7212.101981   8346.992778  ...        0   
264    27955.754000  23296.461670   4160.358618   5540.674809  ...        1   
3151    1075.810833   2151.621667    787.403397    806.182414  ...        1   
5926   25923.072220  38884.608330   4642.237588   4187.146144  ...        1   
7002   22667.190000  37778.650000  18102.749480  14160.168690  ...        0   
8327   10005.648890  10223.306670   4197.653445   2122.712434  ...        0   
11025   3060.546000   3945.698333   3567.134680   3710.983357  ...        0   

       Col2389  Col2390  Col2391  Col2392  Col2393  Col2394  Col2395  Col2396  \
123          0        0        0     29.0     29.0     24.0     12.0     12.0   
264          0        1        2     52.0     52.0     32.0     12.0     12.0   
3151         1        2        2     75.0     75.0     41.0     51.0     51.0   
5926         3        1        3    111.0    111.0     72.0     29.0     29.0   
7002         0        0        0     54.0     54.0     45.0     24.0     24.0   
8327         0        0        0     75.0     70.0     35.0     23.0     19.0   
11025        0        0        0     13.0     13.0      5.0      7.0      7.0   

       Col2397  
123        8.0  
264        7.0  
3151      32.0  
5926      16.0  
7002      20.0  
8327      12.0  
11025      4.0  

[7 rows x 2395 columns]

In [ ]:
data_train[(data_train["Col463"].isna()) & (data_train["Col2"]==1)][["Col2","Col463"]]

# Creating  a data set with both 1 and 0 class as same weightage

In [3]:
na_droped_train = data_train.dropna(axis=1)

In [4]:
na_droped_train.shape

(17521, 1636)

In [5]:
na_droped_train["Col2"].value_counts()

0    15760
1     1761
Name: Col2, dtype: int64

In [8]:
data_train["Col2"].value_counts()

0    15760
1     1761
Name: Col2, dtype: int64

In [6]:
na_droped_train_ones = na_droped_train[na_droped_train["Col2"]==1]

In [7]:
na_droped_train_ones.shape

(1761, 1636)

In [8]:
na_droped_train_zeros = na_droped_train[na_droped_train["Col2"]==0]

In [9]:
na_droped_train_zeros.shape

(15760, 1636)

In [10]:
final_df = na_droped_train_ones.append(na_droped_train_zeros.sample(n=7880))

In [11]:
final_df["Col2"].value_counts()

0    7880
1    1761
Name: Col2, dtype: int64

In [66]:
clf1 = DecisionTreeClassifier(random_state=21,class_weight="balanced")
clf2 = RandomForestClassifier(random_state=21,bootstrap = False, max_features =  10, n_estimators =  1000,class_weight="balanced_subsample")
clf3 = ExtraTreeClassifier(random_state=21,max_features=  2,class_weight="balanced")
clf4 = LogisticRegression(class_weight="balanced")
clf5 = GaussianNB()
clf6 = MLPClassifier(hidden_layer_sizes=(500,500,150,10),activation="relu",solver="adam",learning_rate="adaptive",max_iter=100)
# clf7 = SVC(class_weight="balanced")
clf8 = BernoulliNB()
# clf4 = lgb.train(params, d_train, 100)
# clf4 = LogisticRegression

In [67]:
vclf = VotingClassifier(voting="soft",n_jobs=-1,estimators=[("dt",clf1),("rf",clf2),("et",clf3),("lr",clf4),("gnb",clf5),("mlp",clf6),("bnb",clf8)])

In [14]:
y = final_df["Col2"]
X=final_df.drop(columns=["Col2","Col1"])

In [15]:
print(y.shape,X.shape)

(9641,) (9641, 1634)


In [69]:
vclf.fit(transformed_X,y)

VotingClassifier(estimators=[('dt',
                              DecisionTreeClassifier(class_weight='balanced',
                                                     criterion='gini',
                                                     max_depth=None,
                                                     max_features=None,
                                                     max_leaf_nodes=None,
                                                     min_impurity_decrease=0.0,
                                                     min_impurity_split=None,
                                                     min_samples_leaf=1,
                                                     min_samples_split=2,
                                                     min_weight_fraction_leaf=0.0,
                                                     presort=False,
                                                     random_state=21,
                                                     splitter='best')),
  

In [20]:
vclf.weights

In [21]:
test_data = pd.read_csv("Test.csv")

/home/whirldata/py35/lib/python3.5/site-packages/IPython/core/interactiveshell.py:3044: DtypeWarning:

Columns (700,731,740,752,761,789,811,820,829,841,850) have mixed types. Specify dtype option on import or set low_memory=False.



In [22]:
no_na_test = test_data.dropna(axis=1)

In [23]:
id_cols = list(no_na_test["Col1"])

In [24]:
tets = no_na_test.drop(columns="Col1")

In [70]:
Y_pred = list(vclf.predict(transformed_X_test))
# Y_pred = list(clf.predict(tets))

#### First 1 class and 0 class count with voting classfier 1=7558, 0=12884
#### Second 1 class and 0 class count with voting classfier 1=1959, 0=18483
#### Third this time voting classifier is used with 8 classifiers and dataset has 0 class count = 7880, 1 class count = 1761 [1=1327,0=19115]
#### LDA with same above train data gives 0=1511,1=18931
#### voting classifier after standard scalling  1=2863 and 0= 17579

In [71]:
Y_pred.count(1)

2863

In [72]:
Y_pred.count(0)

17579

In [73]:
final_sub= pd.DataFrame(columns=["Col1","Col2"])

final_sub["Col1"]=id_cols

final_sub["Col2"]=Y_pred

In [75]:
final_sub.to_csv("Submission5_ld.csv",index=False)

In [86]:
print("Using voting classfier gives = ",71.36949)

Using voting classfier gives =  71.36949


In [ ]:
print("Using voting classifier with grid search params tuning = ",)

In [79]:
import lightgbm as lgb
d_train = lgb.Dataset(transformed_X, label=y)
params = {}
params['learning_rate'] = 0.003
params['boosting_type'] = 'gbdt'
params['objective'] = 'binary'
params['metric'] = 'binary_logloss'
params['sub_feature'] = 0.5
params['num_leaves'] = 10
params['min_data'] = 50
params['max_depth'] = 10
clf = lgb.train(params, d_train, 100)


#params
# params = {}
# params['max_bin'] = 10
# params['learning_rate'] = LEARNING_RATE # shrinkage_rate
# params['boosting_type'] = 'gbdt'
# params['objective'] = 'binary'
# params['metric'] = 'l1'          # or 'mae'
# params['sub_feature'] = 0.50      # feature_fraction 
# params['bagging_fraction'] = 0.85 # sub_row
# params['bagging_freq'] = 40
# params['num_leaves'] = 512        # num_leaf
# params['min_data'] = 500         # min_data_in_leaf
# params['min_hessian'] = 0.05     # min_sum_hessian_in_leaf
# params['verbose'] = 0
# clf = lgb.train(params, d_train, 100)

# N_FOLDS = 5
# MAX_BOOST_ROUNDS = 700
# LEARNING_RATE = .0022

# lgb_cv = lgb.cv(params, d_train, num_boost_round=MAX_BOOST_ROUNDS, nfold=N_FOLDS,verbose_eval=20, early_stopping_rounds=40)

# params

# tying grid search cv for random forest

In [97]:
param_grid = [
{'bootstrap': [False], 'n_estimators': [100, 500, 1000, 2000], 'max_features': [2, 3, 4,10,50]},
]

In [98]:
rfc = RandomForestClassifier(random_state=21,n_jobs=-1)

In [99]:
gird_search  = GridSearchCV(rfc,param_grid,scoring="accuracy")

In [100]:
gird_search.fit(X,y)

/home/whirldata/py35/lib/python3.5/site-packages/sklearn/model_selection/_split.py:1978: FutureWarning:

The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.



GridSearchCV(cv='warn', error_score='raise-deprecating',
             estimator=RandomForestClassifier(bootstrap=True, class_weight=None,
                                              criterion='gini', max_depth=None,
                                              max_features='auto',
                                              max_leaf_nodes=None,
                                              min_impurity_decrease=0.0,
                                              min_impurity_split=None,
                                              min_samples_leaf=1,
                                              min_samples_split=2,
                                              min_weight_fraction_leaf=0.0,
                                              n_estimators='warn', n_jobs=-1,
                                              oob_score=False, random_state=21,
                                              verbose=0, warm_start=False),
             iid='warn', n_jobs=None,
             param_grid

In [101]:
gird_search.best_params_

{'bootstrap': False, 'max_features': 10, 'n_estimators': 1000}

In [94]:
gird_search.best_score_

0.7405709214613223

In [95]:
gird_search.scoring

'accuracy'

In [110]:
para_grid = [
{'max_features': [2, 3, 4,10,50]},
]
etc = ExtraTreeClassifier(random_state=21)
grid_search  = GridSearchCV(etc,para_grid,scoring="accuracy")

In [111]:
grid_search.fit(X,y)

/home/whirldata/py35/lib/python3.5/site-packages/sklearn/model_selection/_split.py:1978: FutureWarning:

The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.



GridSearchCV(cv='warn', error_score='raise-deprecating',
             estimator=ExtraTreeClassifier(class_weight=None, criterion='gini',
                                           max_depth=None, max_features='auto',
                                           max_leaf_nodes=None,
                                           min_impurity_decrease=0.0,
                                           min_impurity_split=None,
                                           min_samples_leaf=1,
                                           min_samples_split=2,
                                           min_weight_fraction_leaf=0.0,
                                           random_state=21, splitter='random'),
             iid='warn', n_jobs=None,
             param_grid=[{'max_features': [2, 3, 4, 10, 50]}],
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring='accuracy', verbose=0)

In [112]:
grid_search.best_score_

0.6494601390326875

In [116]:
grid_search.best_params_

{'max_features': 2}

# LDA

In [31]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA

In [42]:
sc =StandardScaler()

In [43]:
transformed_X =  sc.fit_transform(X)

In [44]:
clf =LDA()

In [45]:
clf.fit(X,y)

/home/whirldata/py35/lib/python3.5/site-packages/sklearn/discriminant_analysis.py:388: UserWarning:

Variables are collinear.



LinearDiscriminantAnalysis(n_components=None, priors=None, shrinkage=None,
                           solver='svd', store_covariance=False, tol=0.0001)

In [56]:
transformed_X_test = sc.transform(tets)